### PR TITLE
Static error: fix static error for list of classes

### PR DESCRIPTION
**Pull request by rxng8**
Refer to issue #2798.
list of the errors and warnings (List taken from #2798 ) -
 - [x] mlpack/src/mlpack/methods/adaboost/adaboost_model.hpp 26 warn V690
The 'AdaBoostModel' class implements a move constructor but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/methods/adaboost/adaboost_model.cpp 74 warn V794
The assignment operator should be protected from the case of 'this == &other'.

 - [x] mlpack/src/mlpack/core/tree/hrectbound.hpp 54 warn V690
The 'HRectBound' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/core/tree/ballbound.hpp 32 warn V690
The 'BallBound' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/core/tree/hollow_ball_bound.hpp 33 warn V690
The 'HollowBallBound' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/core/tree/hollow_ball_bound_impl.hpp 82 warn V794
The assignment operator should be protected from the case of 'this == &other'.

 - [x] mlpack/src/mlpack/core/tree/cellbound.hpp 75 warn V690
The 'CellBound' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/core/tree/rectangle_tree/r_tree_split_impl.hpp 441 err V758
The 'range' reference becomes invalid when temporary object returned by a function is destroyed.

 - [x] mlpack/src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp 29 warn V690
The 'DiscreteHilbertValue' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/methods/amf/update_rules/svd_incomplete_incremental_learning.hpp 53 err V730
Not all members of a class are initialized inside the constructor. Consider inspecting: currentUserIndex.

 - [x] mlpack/src/mlpack/methods/amf/update_rules/svd_complete_incremental_learning.hpp 56 err V730
Not all members of a class are initialized inside the constructor. Consider inspecting: currentUserIndex, currentItemIndex.

 - [x] mlpack/src/mlpack/methods/amf/update_rules/svd_complete_incremental_learning.hpp 172 err V730
Not all members of a class are initialized inside the constructor. Consider inspecting: n, m, isStart.

 - [x] mlpack/src/mlpack/methods/amf/termination_policies/simple_residue_termination.hpp 42 warn V730
Not all members of a class are initialized inside the constructor. Consider inspecting: residue, iteration, normOld, nm.

 - [x] mlpack/src/mlpack/methods/range_search/range_search.hpp 42 warn V690
The 'RangeSearch' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/methods/range_search/range_search_impl.hpp 261 warn V1004
The 'referenceTree' pointer was used unsafely after it was verified against nullptr. Check lines: 257, 261.

 - [x] mlpack/src/mlpack/methods/fastmks/fastmks.hpp 63 warn V690
The 'FastMKS' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/methods/fastmks/fastmks_model.hpp 34 warn V690
The 'FastMKSModel' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/methods/fastmks/fastmks_model.cpp 73 warn V794
The assignment operator should be protected from the case of 'this == &other'.

 - [x] mlpack/src/mlpack/methods/hmm/hmm_model.hpp 33 warn V690
The 'HMMModel' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/methods/hoeffding_trees/hoeffding_tree.hpp 61 warn V690
The 'HoeffdingTree' class implements a copy constructor, but lacks the copy assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/methods/hoeffding_trees/hoeffding_tree_model.cpp 64 warn V794
The assignment operator should be protected from the case of 'this == &other'.

 - [x] mlpack/src/mlpack/methods/hoeffding_trees/hoeffding_tree_model.cpp 92 warn V794
The assignment operator should be protected from the case of 'this == &other'.

 - [x] mlpack/src/mlpack/methods/kde/kde.hpp 88 warn V690
The 'KDE' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [x] mlpack/src/mlpack/methods/kmeans/dual_tree_kmeans_rules_impl.hpp 160 warn V519
The 'adjustedScore' variable is assigned values twice successively. Perhaps this is a mistake. Check lines: 159, 160.

 - [x] mlpack/src/mlpack/methods/kmeans/dual_tree_kmeans_rules_impl.hpp 21 err V730
Not all members of a class are initialized inside the constructor. Consider inspecting: lastBaseCase.

 - [ ] mlpack/src/mlpack/methods/lars/lars.cpp 420 warn V688
The 'activeSet' local variable possesses the same name as one of the class members, which can result in a confusion.

 - [x] mlpack/src/mlpack/methods/preprocess/scaling_model.hpp 29 warn V690
The 'ScalingModel' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.

 - [ ] mlpack/src/mlpack/methods/ann/layer/reparametrization_impl.hpp 68 err V562
It's odd to compare a bool type value with a value of double type.

 - [x] mlpack/src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp 23 warn V730
Not all members of a class are initialized inside the constructor. Consider inspecting: reward, deterministic.

 - [x] mlpack/src/mlpack/methods/ann/layer/recurrent_attention_impl.hpp 30 err V730
Not all members of a class are initialized inside the constructor. Consider inspecting: outSize.

 - [x] mlpack/src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp 56 warn V730
Not all members of a class are initialized inside the constructor. Consider inspecting: atomSize, vMin, vMax.

 - [x] mlpack/src/mlpack/methods/amf/termination_policies/incomplete_incremental_termination.hpp 36 warn V730
Not all members of a class are initialized inside the constructor. Consider inspecting: incrementalIndex, iteration.

 - [x] mlpack/src/mlpack/methods/amf/termination_policies/complete_incremental_termination.hpp 37 warn V730
Not all members of a class are initialized inside the constructor. Consider inspecting: incrementalIndex, iteration.

### CODE REVIEW

**Comment by rxng8** [comment]
> Refer to issue #2798.
> list of the errors and warnings -
> 
> * [x]  mlpack/src/mlpack/methods/adaboost/adaboost_model.hpp 26 warn V690
>   The 'AdaBoostModel' class implements a move constructor but lacks the move assignment operator. It is dangerous to use such a class.
> * [x]  mlpack/src/mlpack/methods/adaboost/adaboost_model.cpp 74 warn V794
>   The assignment operator should be protected from the case of 'this == &other'.
> * [x]  mlpack/src/mlpack/core/tree/hrectbound.hpp 54 warn V690
>   The 'HRectBound' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [x]  mlpack/src/mlpack/core/tree/ballbound.hpp 32 warn V690
>   The 'BallBound' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [x]  mlpack/src/mlpack/core/tree/hollow_ball_bound.hpp 33 warn V690
>   The 'HollowBallBound' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [x]  mlpack/src/mlpack/core/tree/hollow_ball_bound_impl.hpp 82 warn V794
>   The assignment operator should be protected from the case of 'this == &other'.
> * [x]  mlpack/src/mlpack/core/tree/cellbound.hpp 75 warn V690
>   The 'CellBound' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [ ]  mlpack/src/mlpack/core/tree/rectangle_tree/r_tree_split_impl.hpp 441 err V758
>   The 'range' reference becomes invalid when temporary object returned by a function is destroyed.
> * [ ]  mlpack/src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp 29 warn V690
>   The 'DiscreteHilbertValue' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [ ]  mlpack/src/mlpack/methods/amf/update_rules/svd_incomplete_incremental_learning.hpp 53 err V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: currentUserIndex.
> * [ ]  mlpack/src/mlpack/methods/amf/update_rules/svd_complete_incremental_learning.hpp 56 err V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: currentUserIndex, currentItemIndex.
> * [ ]  mlpack/src/mlpack/methods/amf/update_rules/svd_complete_incremental_learning.hpp 172 err V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: n, m, isStart.
> * [ ]  mlpack/src/mlpack/methods/amf/termination_policies/simple_residue_termination.hpp 42 warn V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: residue, iteration, normOld, nm.
> * [ ]  mlpack/src/mlpack/methods/range_search/range_search.hpp 42 warn V690
>   The 'RangeSearch' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [ ]  mlpack/src/mlpack/methods/range_search/range_search_impl.hpp 261 warn V1004
>   The 'referenceTree' pointer was used unsafely after it was verified against nullptr. Check lines: 257, 261.
> * [ ]  mlpack/src/mlpack/methods/fastmks/fastmks.hpp 63 warn V690
>   The 'FastMKS' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [ ]  mlpack/src/mlpack/methods/fastmks/fastmks_model.hpp 34 warn V690
>   The 'FastMKSModel' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [ ]  mlpack/src/mlpack/methods/fastmks/fastmks_model.cpp 73 warn V794
>   The assignment operator should be protected from the case of 'this == &other'.
> * [ ]  mlpack/src/mlpack/methods/hmm/hmm_model.hpp 33 warn V690
>   The 'HMMModel' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [ ]  mlpack/src/mlpack/methods/hoeffding_trees/hoeffding_tree.hpp 61 warn V690
>   The 'HoeffdingTree' class implements a copy constructor, but lacks the copy assignment operator. It is dangerous to use such a class.
> * [ ]  mlpack/src/mlpack/methods/hoeffding_trees/hoeffding_tree_model.cpp 64 warn V794
>   The assignment operator should be protected from the case of 'this == &other'.
> * [ ]  mlpack/src/mlpack/methods/hoeffding_trees/hoeffding_tree_model.cpp 92 warn V794
>   The assignment operator should be protected from the case of 'this == &other'.
> * [ ]  mlpack/src/mlpack/methods/kde/kde.hpp 88 warn V690
>   The 'KDE' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [ ]  mlpack/src/mlpack/methods/kmeans/dual_tree_kmeans_rules_impl.hpp 160 warn V519
>   The 'adjustedScore' variable is assigned values twice successively. Perhaps this is a mistake. Check lines: 159, 160.
> * [ ]  mlpack/src/mlpack/methods/kmeans/dual_tree_kmeans_rules_impl.hpp 21 err V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: lastBaseCase.
> * [ ]  mlpack/src/mlpack/methods/lars/lars.cpp 420 warn V688
>   The 'activeSet' local variable possesses the same name as one of the class members, which can result in a confusion.
> * [ ]  mlpack/src/mlpack/methods/preprocess/scaling_model.hpp 29 warn V690
>   The 'ScalingModel' class implements a move constructor, but lacks the move assignment operator. It is dangerous to use such a class.
> * [x]  mlpack/src/mlpack/methods/ann/layer/reparametrization_impl.hpp 68 err V562
>   It's odd to compare a bool type value with a value of double type.
> * [ ]  mlpack/src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp 23 warn V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: reward, deterministic.
> * [ ]  mlpack/src/mlpack/methods/ann/layer/recurrent_attention_impl.hpp 30 err V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: outSize.
> * [ ]  mlpack/src/mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp 56 warn V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: atomSize, vMin, vMax.
> * [ ]  mlpack/src/mlpack/methods/amf/termination_policies/incomplete_incremental_termination.hpp 36 warn V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: incrementalIndex, iteration.
> * [ ]  mlpack/src/mlpack/methods/amf/termination_policies/complete_incremental_termination.hpp 37 warn V730
>   Not all members of a class are initialized inside the constructor. Consider inspecting: incrementalIndex, iteration.

**Comment by rxng8** [comment]
@rcurtin Hi, as I am checking the static error, I encounter an error about calculating the `adjustedScore` in [`dual_tree_kmeans_rules_impl.hpp`](https://github.com/mlpack/mlpack/blob/94dc6be5b5324158e2fbaed70e6148f87a22930f/src/mlpack/methods/kmeans/dual_tree_kmeans_rules_impl.hpp#L159), the error said "The 'adjustedScore' variable is assigned values twice successively. Perhaps this is a mistake. Check lines: 160, 161." [Check here](http://ci.mlpack.org/job/pull-requests-mlpack-static-code-analysis/7610/cppcheck/folder.-1265198708/source.9d7b07dd-b4a2-4484-80eb-9880941136be/#161)

Basically, the adjusted score is computed 2 times:
```
adjustedScore = lastScore + lastQueryDescDist;
adjustedScore = lastScore + lastRefDescDist;
```
In the comment you said that "the last score is equal to the distance between the centroids minus the radii of the query and reference bounds along the axis of the line between the two centroids." So can I fix it like this?
```
adjustedScore = lastScore + lastQueryDescDist + lastRefDescDist;
```

**Comment by rcurtin** [comment]
@rxng8 I think your suggestion is correct.  It looks like the current implementation is definitely a bug.  If you change to the code you proposed, do all the tests still pass?

**Comment by rxng8** [comment]
@rcurtin Yes, all the `[KMeansTest]` still passed!

**Comment by rcurtin** [comment]
@rxng8 awesome!  Then I think that is the correct fix. :+1:

File: src/mlpack/core/tree/hrectbound_impl.hpp [hunk 1]
```diff
@@ -103,6 +103,27 @@ inline HRectBound<MetricType, ElemType>::HRectBound(
   other.minWidth = 0.0;
 }
 
+/**
+ * Move assignment operator
```
Comment (zoq) [inline src/mlpack/core/tree/hrectbound_impl.hpp:107]:
```suggestion
 * Move assignment operator.
```

File: src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp [hunk 2]
```diff
@@ -182,6 +182,14 @@ class DiscreteHilbertValue
    */
   DiscreteHilbertValue& operator=(const DiscreteHilbertValue& val);
 
+   /**
```
Comment (zoq) [inline src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp:185]:
```suggestion
  /**
```

File: src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp [hunk 3]
```diff
@@ -182,6 +182,14 @@ class DiscreteHilbertValue
...
+   * Move the local Hilbert object.
+   *
+   * @param val The DiscreteHilbertValue object from which the dataset
+   *    will be copied.
+   */
+  DiscreteHilbertValue& operator=(DiscreteHilbertValue&& val);
```
Comment (zoq) [inline src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp:191]: I would use `other` instead of `val` here as well, it's what you used in the implementation as well.

File: src/mlpack/methods/range_search/range_search_impl.hpp [hunk 4]
```diff
@@ -254,12 +290,15 @@ void RangeSearch<MetricType, MatType, TreeType>::Train(
     throw std::invalid_argument("cannot train on given reference tree when "
         "naive search (without trees) is desired");
 
+  // Can only train when passed argument `referenceTree` is not nullptr
```
Comment (zoq) [inline src/mlpack/methods/range_search/range_search_impl.hpp:293]:
```suggestion
  // Can only train when passed argument `referenceTree` is not nullptr.
```

File: src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp (hunk 3 above)
Comment (rxng8) [reply src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp:191]: sure

**Comment by rxng8** [comment]
@zoq Hi, I have finished the fix. Please feel free to review it!

**Comment by zoq** [comment]
Looks like there are two issues that aren't checked yet, but I guess you checked everything?

**Comment by rxng8** [comment]
@zoq As you can see from [this static code check](https://gauravghati.github.io/StaticCodeAnalysisReport/) (from @gauravghati, issue #2798 ) which includes the [`reparametrization` static error](https://gauravghati.github.io/StaticCodeAnalysisReport/sources/reparametrization_impl.hpp_45.html#ln68) and the  [`lars` static error](https://gauravghati.github.io/StaticCodeAnalysisReport/sources/lars.cpp_28.html#ln420). 
* For the `reparametrization` static error, there is no comparison in the error line (and also no parenthesis), so we cannot conclude what happened there. (Maybe it is a bug of the static code checker?)
* for the `lars` static error, the class variable `activeSet` is intended to be there to be modified, there are no other method parameters called `activeSet` as well as predefined variable `activeSet` or assignment of the `activeSet` variable.

Therefore, I believe that it is fine now, but I will still leave that unchecked!

**Comment by zoq** [comment]
> @zoq As you can see from [this static code check](https://gauravghati.github.io/StaticCodeAnalysisReport/) (from @gauravghati, issue #2798 ) which includes the [`reparametrization` static error](https://gauravghati.github.io/StaticCodeAnalysisReport/sources/reparametrization_impl.hpp_45.html#ln68) and the [`lars` static error](https://gauravghati.github.io/StaticCodeAnalysisReport/sources/lars.cpp_28.html#ln420).
> 
>     * For the `reparametrization` static error, there is no comparison in the error line (and also no parenthesis), so we cannot conclude what happened there. (Maybe it is a bug of the static code checker?)
> 
>     * for the `lars` static error, the class variable `activeSet` is intended to be there to be modified, there are no other method parameters called `activeSet` as well as predefined variable `activeSet` or assignment of the `activeSet` variable. So everything is fine! (Maybe it is a bug of the static code checker?)
> 
> 
> Therefore, I believe that it is fine now, but I will still leave those unchecked!

Okay, perfect, I looked at the two issues and I think they are false postives, so safe to ignore at this point.

File: src/mlpack/core/tree/hrectbound.hpp [hunk 5]
```diff
@@ -73,12 +73,16 @@ class HRectBound
...
   //! Copy constructor; necessary to prevent memory leaks.
   HRectBound(const HRectBound& other);
+
   //! Same as copy constructor; necessary to prevent memory leaks.
   HRectBound& operator=(const HRectBound& other);
 
   //! Move constructor: take possession of another bound's information.
   HRectBound(HRectBound&& other);
 
+  //! Move assignment operator
```
Comment (zoq) [inline src/mlpack/core/tree/hrectbound.hpp:83]:
```suggestion
  //! Move assignment operator.
```

File: src/mlpack/methods/amf/termination_policies/complete_incremental_termination.hpp [hunk 6]
```diff
@@ -36,7 +36,8 @@ class CompleteIncrementalTermination
    */
   CompleteIncrementalTermination(
       TerminationPolicy tPolicy = TerminationPolicy()) :
-      tPolicy(tPolicy) { }
+      tPolicy(tPolicy), incrementalIndex(0), iteration(0)
+  { /** Nothing to do here. */ }
```
Comment (zoq) [inline src/mlpack/methods/amf/termination_policies/complete_incremental_termination.hpp:40]:
```suggestion
  { /* Nothing to do here. */ }
```

File: src/mlpack/methods/amf/termination_policies/incomplete_incremental_termination.hpp [hunk 7]
```diff
@@ -35,7 +35,8 @@ class IncompleteIncrementalTermination
    */
   IncompleteIncrementalTermination(
       TerminationPolicy tPolicy = TerminationPolicy()) :
-      tPolicy(tPolicy) { }
+      tPolicy(tPolicy), incrementalIndex(0), iteration(0)
+  { /** Nothing to do here. */ }
```
Comment (zoq) [inline src/mlpack/methods/amf/termination_policies/incomplete_incremental_termination.hpp:39]:
```suggestion
  { /* Nothing to do here. */ }
```

**Review by zoq — COMMENTED** [review]
Just some really minor style issues; the rest looks good.

**Comment by rxng8** [comment]
@zoq I have committed your suggested changes! Thanks!

**Review by zoq — APPROVED** [review]
Thanks for working on this one, no more comments from my side.

File: src/mlpack/core/tree/ballbound_impl.hpp [hunk 8]
```diff
@@ -92,6 +92,22 @@ BallBound<MetricType, VecType>::BallBound(BallBound&& other) :
   other.ownsMetric = false;
 }
 
+//! Move assignment operator.
+template<typename MetricType, typename VecType>
+BallBound<MetricType, VecType>& BallBound<MetricType, VecType>::operator=(
+    BallBound&& other)
+{
```
Comment (rcurtin) [inline src/mlpack/core/tree/ballbound_impl.hpp:99]: Just a little note, I think we need to add a guard here in case `this` is `other`:

```
if (&other != this)
```

Also, don't forget the `return *this` at the end. :)

I realize the copy constructor has these issues too---do you think you could also fix those?

File: src/mlpack/core/tree/hrectbound_impl.hpp [hunk 9]
```diff
@@ -103,6 +103,27 @@ inline HRectBound<MetricType, ElemType>::HRectBound(
...
 }
 
+/**
+ * Move assignment operator.
+ */
+template<typename MetricType, typename ElemType>
+inline HRectBound<
+    MetricType,
+    ElemType>& HRectBound<MetricType,
+    ElemType>::operator=(HRectBound<MetricType, ElemType>&& other)
```
Comment (rcurtin) [inline src/mlpack/core/tree/hrectbound_impl.hpp:113]:
```suggestion
inline HRectBound<MetricType, ElemType>&
HRectBound<MetricType, ElemType>::operator=(
    HRectBound<MetricType, ElemType>&& other)
```
Just a different suggestion for wrapping; I think it's a bit cleaner, but up to you. :+1:

File: src/mlpack/methods/amf/termination_policies/simple_residue_termination.hpp [hunk 10]
```diff
@@ -41,7 +41,15 @@ class SimpleResidueTermination
    */
   SimpleResidueTermination(const double minResidue = 1e-5,
                            const size_t maxIterations = 10000)
-      : minResidue(minResidue), maxIterations(maxIterations) { }
+      : minResidue(minResidue),
```
Comment (rcurtin) [inline src/mlpack/methods/amf/termination_policies/simple_residue_termination.hpp:44]: GitHub won't let me suggest it, but I think the 'more correct' way to style this according to our style guide is to put the `:` on the previous line (e.g. `= 10000) :`) and then move all these initialization members over two spaces.

File: src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp [hunk 11]
```diff
@@ -34,8 +34,7 @@ void ReinforceNormal<InputDataType, OutputDataType>::Forward(
   if (!deterministic)
   {
     // Multiply by standard deviations and re-center the means to the mean.
-    output = arma::randn<arma::Mat<eT> >(input.n_rows, input.n_cols) *
-        stdev + input;
+    output = output.randn(input.n_rows, input.n_cols) * stdev + input;
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp:37]: I didn't know `randn()` returned the matrix so that it could be used like this.  Nice!

File: src/mlpack/methods/hmm/hmm_model.hpp [hunk 12]
```diff
@@ -129,6 +129,20 @@ class HMMModel
...
+  //! Move assignment operator.
+  HMMModel& operator=(HMMModel&& other)
+  {
+    if (this != &other)
+    {
+      type = other.type;
+      discreteHMM = other.discreteHMM;
+      gaussianHMM = other.gaussianHMM;
+      gmmHMM = other.gmmHMM;
+      diagGMMHMM = other.diagGMMHMM;
```
Comment (rcurtin) [inline src/mlpack/methods/hmm/hmm_model.hpp:141]: Do we need to clear the members of `other`?

File: src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp [hunk 13]
```diff
@@ -224,6 +224,117 @@ HoeffdingTree<FitnessFunction, NumericSplitType, CategoricalSplitType>::
...
+    successProbability(other.successProbability),
+    splitDimension(other.splitDimension),
+    majorityClass(other.majorityClass),
+    majorityProbability(other.majorityProbability),
+    categoricalSplit(std::move(other.categoricalSplit)),
+    numericSplit(std::move(other.numericSplit))
+{
+  // Remove pointers.
+  other.dimensionMappings = nullptr;
+  other.datasetInfo = nullptr;
```
Comment (rcurtin) [inline src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp:253]: Should we set the other members of `other` back to their defaults?  I'm not sure whether it's really necessary.

File: src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp [hunk 14]
```diff
@@ -341,7 +452,7 @@ void HoeffdingTree<
     delete dimensionMappings;
 
   const CategoricalSplitType<FitnessFunction> categoricalSplitIn(0, 0);
-  const NumericSplitType<FitnessFunction>& numericSplitIn(0);
+  const NumericSplitType<FitnessFunction> numericSplitIn(0);
```
Comment (rcurtin) [inline src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp:455]: Wow, did this even compile before?  Nice fix!

File: src/mlpack/methods/range_search/range_search_impl.hpp [hunk 15]
```diff
@@ -254,12 +290,15 @@ void RangeSearch<MetricType, MatType, TreeType>::Train(
     throw std::invalid_argument("cannot train on given reference tree when "
         "naive search (without trees) is desired");
 
+  // Can only train when passed argument `referenceTree` is not nullptr.
```
Comment (rcurtin) [inline src/mlpack/methods/range_search/range_search_impl.hpp:293]: Another nice catch! :+1:

**Review by rcurtin — COMMENTED** [review]
@rxng8 thank you so much for working on all these issues!  It's great to have them all fixed.  I took a look through and I left a few comments, but really only one needs to be addressed before merge---it seems there is a copy/move constructor pair that don't have the `if (this != &other)` guard.  Let me know if I overlooked something. :)

File: src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp (hunk 11 above)
Comment (rxng8) [reply src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp:37]: thank you!

File: src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp (hunk 14 above)
Comment (rxng8) [reply src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp:455]: Thanks haha!

File: src/mlpack/core/tree/ballbound_impl.hpp (hunk 8 above)
Comment (rxng8) [reply src/mlpack/core/tree/ballbound_impl.hpp:99]: sure! I will do!

File: src/mlpack/methods/amf/termination_policies/simple_residue_termination.hpp (hunk 10 above)
Comment (rxng8) [reply src/mlpack/methods/amf/termination_policies/simple_residue_termination.hpp:44]: sure! Will change that!

File: src/mlpack/methods/hmm/hmm_model.hpp (hunk 12 above)
Comment (rxng8) [reply src/mlpack/methods/hmm/hmm_model.hpp:141]: Yes, we do! Thanks for that!

File: src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp (hunk 13 above)
Comment (rxng8) [reply src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp:253]: I think because only two of them are pointers, then we have to set the other's pointers to null. For the other, there are variables that are objects (E.g: `categoricalSplits` is an `std::vector`), they have been converted to rvalue reference by `std::move`, other variables are primary type so they don't need to be set to null.

Comment (rxng8) [reply src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp:253]: Well I think for the primary type variables, I should have reset it to default value! :)

File: src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp (hunk 11 above)
Comment (zoq) [reply src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp:37]: Same here, actually I had to test it out.

Comment (rxng8) [reply src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp:37]: Yes, I also tested it and it seems to return a matrix. did it happen to behave the same ways?

Comment (rxng8) [reply src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp:37]: I believe it is :))

Comment (rcurtin) [reply src/mlpack/methods/ann/layer/reinforce_normal_impl.hpp:37]: Yep, looks great, thank you!
